netmiko/nokia/nokia_srl.py

```diff
@@ -0,0 +1,197 @@
+        """
+        if not pattern:
+            output = self.read_channel_timing(read_timeout=10.0)
+        else:
+            output = self.read_until_pattern(pattern=pattern)
+
+        matches = re.search(check_string, output)
+        return True if matches else False
+
+    def commit(self) -> str:
+        """Commit changes by using 'commit stay'."""
+        cmd = "commit stay"
+        self.write_channel(self.normalize_cmd(cmd))
+        output = (
+            self._get_cmd_output_and_prompt(cmd)
+            if self.global_cmd_verify is not False
+            else ""
+        )
+        return output
+
+    def save_config(
+        self,
+        cmd: str = "save startup",
+        confirm: bool = False,
+        confirm_response: str = "",
+    ) -> str:
+        """Save current running configuration as initial (startup) configuration"""
+        self.write_channel(self.normalize_cmd(cmd))
+        output = (
+            self._get_cmd_output_and_prompt(cmd)
+            if self.global_cmd_verify is not False
+            else ""
+        )
+        return output
+
+    def exit_config_mode(self, exit_config: str = "", pattern: str = "") -> str:
+        """Exit the candidate private mode"""
+        output = ""
+        self.write_channel(self.RETURN)
+        prompt = self.read_until_pattern(pattern="]--")
+        matches = re.search(r"\n--{( | \* | \+ | \+\* | \!\+ | \!\* )candidate", prompt)
+        if matches:
+            # In config mode and changes were made"""
+            # Get the subgroup in matches. Should be only one.
+            group = matches.group()
+            if "*" in group:
+                # Changes were made but not committed. Discarding changes
+                output += self._discard()
+            elif "+" in group:
+                # Changes were made, committed and discarding is not available.
+                # Committed changes can be saved in 'running mode'.
+                log.warning(
+                    """Exiting candidate private mode with unsaved changes!
+                    Changes can be saved in running mode."""
+                )
+        # Switch to 'running' mode
+        output += self._running_mode()
+        return output
+
+    def send_config_set(
+        self,
+        config_commands: Union[str, Sequence[str], TextIO, None] = None,
+        exit_config_mode: bool = False,
+        **kwargs: Any,
+    ) -> str:
+        """Nokia SRL requires you not exit from configuration mode."""
+        return super().send_config_set(
+            config_commands=config_commands, exit_config_mode=exit_config_mode, **kwargs
+        )
+
+    def _discard(self) -> str:
+        """Discard changes made in candidate private mode"""
+        log.warning("Uncommitted changes will be discarted!")
+        cmd = "discard stay"
+        self.write_channel(self.normalize_cmd(cmd))
+        output = (
+            self._get_cmd_output_and_prompt(cmd)
```

Comment:
Once again pretty strong preference just to use standard Netmiko practice of calling `self._send_command_str(cmd)`.